Wasp A La Carte

beardeddude opened the pull request:
Wasp A La Carte is a build system enhancement that allows the wasp-os build to be configured via a simple toml file.

Features:

- A simple centralized human readable config file
- Easy selection of frozen apps, watch faces and the quick ring
- Automatic management of the software and faces apps
- No confusion for where to place apps for new devs
- Consistent naming across all apps
- A slim default build (less needs to be shown off)
- A more flexible path for future growth

What I changed:

- Moved all but the essential system apps to the root app directory
- Renamed app files or classes as needed to make them all consistent
- Moved all watch faces to a watch face directory
- Split up the manifests
- Added a in-os app registry
- Created a config script in python and integrated it into the Makefile
- Updated the documentation

Tested the following:
- Simulator
- Pinetime
- Various configurations of the toml besides the one submitted
- Loading frozen and flash apps
- Building via: local, docker, github workflows

Notes on the code:

- There is a bug between pip, docker and the tomli package that results in tomli not being installed correctly within the docker image build. This bug is not present on local or workflows builds. As a result I had to put a pip install command in the make file to ensure the package was installed at compile time
- The default build in the toml is probably overly slim but is based on the existing default build. Its not my place to pick the defaults so please change to your liking. I aimed this toml mostly at showing what options were available within it.
- The steps app was the only ring app not moved to the toml. This is because it has special integrations in wasp.py and this PR is already huge. I think it could be modifed to be able to drop in and out without issue but thats for another PR if this one is adopted. For now is pretty easy to follow how its hard coded (added to the end of the quick ring). Ideally I think it should be removable.
- configure_wasp_apps.py is probably a little heavy on string manipulation. I bet there is some clever way to just create and print out the objects but I don't do much work in python and went with what i know would work.

Ideas for the future if this is adopted:

- Move steps app
- Automatically strip icons out of apps placed on the quick ring if they have them to save space
- Specify theme in the toml


Signed-off-by: Adam Blair adampblair@protonmail.com

daniel-thompson commented:
Overall this looks like a great idea. A really good way to cope with the limited FLASH on nRF52832.

However I noticed that these changes are preventing the test autodiscovery from working correctly. Below is a diff of the output logs of a latest mainline versus what appears in the github CI report (lines starting `-` show test that don't appear in the github CI report). It shows 60 tests are no longer being discovered.

~~~
 wasp/boards/simulator/test_qa.py::test_app_library[AlarmApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[BeaconApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[CalculatorApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[ChronoApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[ClockApp0] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[ClockApp1] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[DemoApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[DisaBLEApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[DualClockApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[FacesApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[FibonacciClockApp] PASSED
+wasp/boards/simulator/test_qa.py::test_app_library[FlashlightApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[GalleryApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[GameOfLifeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[HaikuApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[HeartApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[LauncherApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[LevelApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[MorseApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[MusicPlayerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[NotificationApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[Play2048App] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[ReadMeApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[SettingsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[SnakeGameApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[SoftwareApp] PASSED
+wasp/boards/simulator/test_qa.py::test_app_library[SnakeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[SportsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[StepCounterApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[StopwatchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[TemplateApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[TestApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[TimerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[TorchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_app_library[WeatherApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[WeekClockApp] PASSED
-wasp/boards/simulator/test_qa.py::test_app_library[WordClockApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[AlarmApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[BeaconApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[CalculatorApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[ChronoApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[ClockApp0] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[ClockApp1] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[DemoApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[DisaBLEApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[DualClockApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[FacesApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[FibonacciClockApp] PASSED
+wasp/boards/simulator/test_qa.py::test_docstrings[FlashlightApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[GalleryApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[GameOfLifeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[HaikuApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[HeartApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[LauncherApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[LevelApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[MorseApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[MusicPlayerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[NotificationApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[Play2048App] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[ReadMeApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[SettingsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[SnakeGameApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[SoftwareApp] PASSED
+wasp/boards/simulator/test_qa.py::test_docstrings[SnakeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[SportsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[StepCounterApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[StopwatchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[TemplateApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[TestApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[TimerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[TorchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_docstrings[WeatherApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[WeekClockApp] PASSED
-wasp/boards/simulator/test_qa.py::test_docstrings[WordClockApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[AlarmApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[BeaconApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[CalculatorApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[ChronoApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[ClockApp0] PASSED
-wasp/boards/simulator/test_qa.py::test_README[ClockApp1] PASSED
 wasp/boards/simulator/test_qa.py::test_README[DemoApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[DisaBLEApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[DualClockApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[FacesApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[FibonacciClockApp] PASSED
+wasp/boards/simulator/test_qa.py::test_README[FlashlightApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[GalleryApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[GameOfLifeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[HaikuApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[HeartApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[LauncherApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[LevelApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[MorseApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[MusicPlayerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[NotificationApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[Play2048App] PASSED
 wasp/boards/simulator/test_qa.py::test_README[ReadMeApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[SettingsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[SnakeGameApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[SoftwareApp] PASSED
+wasp/boards/simulator/test_qa.py::test_README[SnakeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[SportsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[StepCounterApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[StopwatchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[TemplateApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[TestApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[TimerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[TorchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_README[WeatherApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[WeekClockApp] PASSED
-wasp/boards/simulator/test_qa.py::test_README[WordClockApp] PASSED
+wasp/boards/simulator/test_qa.py::test_screenshot[AlarmApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[BeaconApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[CalculatorApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[ChronoApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[ClockApp0] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[ClockApp1] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[DemoApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[DisaBLEApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[DualClockApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[FacesApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[FibonacciClockApp] PASSED
+wasp/boards/simulator/test_qa.py::test_screenshot[FlashlightApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[GalleryApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[GameOfLifeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[HaikuApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[HeartApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[LauncherApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[LevelApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[MorseApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[MusicPlayerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[NotificationApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[Play2048App] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[ReadMeApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[SettingsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[SnakeGameApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[SoftwareApp] PASSED
+wasp/boards/simulator/test_qa.py::test_screenshot[SnakeApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[SportsApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[StepCounterApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[StopwatchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[TemplateApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[TestApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[TimerApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[TorchApp] PASSED
 wasp/boards/simulator/test_qa.py::test_screenshot[WeatherApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[WeekClockApp] PASSED
-wasp/boards/simulator/test_qa.py::test_screenshot[WordClockApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_app[Heart] PASSED
 wasp/boards/simulator/test_smoke.py::test_app[Settings] PASSED
 wasp/boards/simulator/test_smoke.py::test_app[Software] PASSED
@@ -145,38 +98,26 @@
 wasp/boards/simulator/test_smoke.py::test_constructor[AlarmApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[BeaconApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[CalculatorApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[ChronoApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[ClockApp0] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[ClockApp1] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[DemoApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[DisaBLEApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[DualClockApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[FacesApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[FibonacciClockApp] PASSED
+wasp/boards/simulator/test_smoke.py::test_constructor[FlashlightApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[GalleryApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[GameOfLifeApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[HaikuApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[HeartApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[LauncherApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[LevelApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[MorseApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[MusicPlayerApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[NotificationApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[Play2048App] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[ReadMeApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[SettingsApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[SnakeGameApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[SoftwareApp] PASSED
+wasp/boards/simulator/test_smoke.py::test_constructor[SnakeApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[SportsApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[StepCounterApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[StopwatchApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[TemplateApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[TestApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[TimerApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[TorchApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_constructor[WeatherApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[WeekClockApp] PASSED
-wasp/boards/simulator/test_smoke.py::test_constructor[WordClockApp] PASSED
 wasp/boards/simulator/test_smoke.py::test_launcher_ring PASSED
 wasp/boards/simulator/test_smoke.py::test_quick_ring PASSED
 wasp/boards/simulator/test_smoke.py::test_selftest_crash PASSED
~~~

beardeddude commented:
Good call out. I missed that. It looks like there are 3 reasons for this I see in that diff.

1. I forgot to add the watch faces dir to the auto discovery
2. I moved the system apps to a sub dir but didnt add that to auto discovery
3. I renamed some of the app classes for consistency so the class name changed (ie: TorchApp -> FlashlightApp) but the app is still there it just shows up in the diff because of that.

I'll get that fixed and clear the conflicts with the other PRs u just merged.

beardeddude commented:
Updated and ready for review. 

While merging the conflicts and updating the tests I realized there was still a lot of opportunity for more consistency and organization. While this is a lot of refactoring I have noticed that a LOT of new devs seem to get stuck on these points and because this update is touching these areas anyways I just decided to go all the way. Because of the naming relationship between the class and the module I think it makes sense to have all the naming requirements be derivations of the class name. That way the NAME variable is the only exception and only for the reason it needs to exist in the first place (to be a short name for the launcher).

Changes:

- Added watch_faces and system apps dirs to the auto discovery
- Added a new qa test to enforce the module to class naming convention required by the software app
- Changed the existing qa tests to be based on class name not the NAME var
- Created screenshot and icon res sub dirs and reorganized the res files
- Renamed the screenshots to be based on the class name of the app
- Renamed the icon files to be based on the class name of the app
- Documented the naming conventions in detail with an example
- Specifically pointed new devs to "make check"
- Updated the README.rst to reflect the configurable nature of this build system
- Stopped internal imports from causing duplicates in the tests (clock app)

Notes:
I did not add a test to enforce the existence and naming of the icon files in res. Many apps don't have one, or don't have an icon at all and none are really required to build. So I just renamed them to all be consistent and made sure the documentation strongly encouraged compliance.

daniel-thompson commented on Makefile:
```diff
@@ -73,6 +76,15 @@ micropython: build-$(BOARD_SAFE) wasp/boards/$(BOARD_SAFE)/watch.py micropython/
 		build-$(BOARD)/micropython.zip
 
+wasp/boards/manifest_user_apps.py: wasp.toml
+	pip3 install tomli
```
I don't think its appropriate to install packages from a build rule.

beardeddude replied:
I 100% agree but there is a bug between pip, docker and the tomli package that results in tomli not being installed correctly within the docker image build. This bug is not present on local or workflows builds. As a result I had to put a pip install command in the make file to ensure the package was installed at compile time.

I worked for hours to find another way around this bug.

- I tried the toml package (the other toml lib) and had the same issue. I think one is based on the other. 
- I tried putting the install on a new layer of the docker image
- Tried different orders of the image
- Tried all the docker ubuntu images to try and find one without the bug
- Tried debian images instead of ubuntu

It's possible its related to https://github.com/pypa/setuptools_scm/issues/625 but that bug was fixed a while ago so im not sure.

One solution is to require python 3.11 (which has tomli built in) but that version is pretty bleeding edge for local dev and it would require tomli be imported as tomlib so it puts us in a weird position of either requiring python 3.11 or doing some kind of conditional import. 

I felt this was the least evil solution. Maybe i should put a comment in the makefile tho?

beardeddude replied:
I would be interested if you experience this bug when building via docker with is line removed. Or if you have any ideas on where else this command could be added. To be honest im at a bit of a loss as to how to fix this. Having that install command in the makefile is not great.

I run Docker version 20.10.17, build 100c701 on a Debian based system

daniel-thompson replied:
I have to admit I never use or test docker build code (because it isn't compatible with podman)... so if you had been less conscientious I would never have noticed and it would probably have been left to the docker users to figure this out!

However... I think you can probably work around this by upgrading the Dockerfile to `FROM ubuntu:jammy` and add python3-tomli to the package list (instead of using pip3).

beardeddude replied:
I figured out what was happening. I was running the shell script with sudo which resulted in the docker build command failing but then the docker run command was just using an old image. So i never got my updates. 

Removed that line from the makefile. it wasn't a real bug, just me being dumb lol.

daniel-thompson commented on README.rst:
```diff
@@ -134,138 +140,149 @@ PineTime:
    :width: 233
 
-Screenshots of the built in applications running on the wasp-os
+Screenshots of the available in applications running on the wasp-os
```
Minor typo (no "in")

beardeddude replied:
Fixed.

daniel-thompson commented on README.rst:
```diff
@@ -134,138 +140,149 @@ PineTime:
-.. image:: res/LifeApp.png
-   :alt: Game of Life running in the wasp-os simulator
+.. image:: res/screenshots/DisaBLEApp.png
+   :alt: Small application for disabling bluetooth to save power and enhance security
```
DisaBLE is a bit of a hacky (if we ever figure out how to do this properly we'll move the feature to Settings) but whilst it is with us I think it counts as a system app.

beardeddude replied:
Agreed. Moved.

daniel-thompson commented on Makefile:
```diff
@@ -61,7 +64,7 @@ wasp/boards/$(BOARD_SAFE)/watch.py : wasp/boards/$(BOARD_SAFE)/watch.py.in
 	$(MAKE) -C micropython/mpy-cross
 
-micropython: build-$(BOARD_SAFE) wasp/boards/$(BOARD_SAFE)/watch.py micropython/mpy-cross/mpy-cross
+micropython: build-$(BOARD_SAFE) wasp/boards/manifest_user_apps.py wasp/boards/$(BOARD_SAFE)/watch.py micropython/mpy-cross/mpy-cross
```
I think this new dependency is missing from the simulator (`make sim` currently doesn't work until after you have done a build for the watch).

beardeddude replied:
Great catch. Fixed. Thanks for the thorough review. This PR ended up being about twice as big as I was planning haha

daniel-thompson reviewed (changes requested):
Looks good!

Still haven't tested on the watch but tests and simulator look good.

beardeddude commented:
One last small update. Because of the enforced naming relationship the toml file didn't need both class and file. So I updated it to only need file.

Everything should be addressed and good to go.

fuzzy7k commented:
The Makefile needs attention. If I work on a new watchface, the changes do not get picked up.

beardeddude commented:
The make file only checks the wasp.toml. I did not see a way to check all of the specified apps in a clean way. If you are developing and making changes in the /app or /watchfaces dir but have already built it will continue to use the one it already built with previously. Its not an issue if all you are doing is changing what apps are included. If developing you can touch wasp.toml or do a "make clean". If you have a good solution to this limitation please open a PR.

fuzzy7k commented:
Seems like you should target those folders and use configure_wasp_apps.py as a filter. Maybe break that up into two components, one rule that is responsible for filtering the apps, and another rule to build the manifests & such.

The manifest does not need to be rebuilt if dev work is being done, so that part works. Just make a copy_wasp_apps.py or such that compares the argument with the toml.
